File: source/go/coin_change.go
package main

import "fmt"

func coinChange(coins []int, amount int) int {
	dp := make([]int, amount+1)
	for i := 1; i <= amount; i++ {
		dp[i] = amount + 1
	}
	dp[0] = 0

	for i := 1; i <= amount; i++ {
		for _, coin := range coins {
			if coin <= i {
				candidate := dp[i-coin] + 1
				if candidate < dp[i] {
					dp[i] = candidate
				}
			}
		}
	}

	if dp[amount] > amount {
		return -1
	}
	return dp[amount]
}

func main() {
	coins := []int{1, 5, 10, 25}
	amount := 36
	result := coinChange(coins, amount)
	fmt.Printf("min coins: %d\n", result)
}

File: source/python/coin_change.py
def coin_change(coins, amount):
    dp = [amount + 1] * (amount + 1)
    dp[0] = 0
    for i in range(1, amount + 1):
        for coin in coins:
            if coin <= i:
                candidate = dp[i - coin] + 1
                if candidate < dp[i]:
                    dp[i] = candidate
    return dp[amount] if dp[amount] <= amount else -1

coins = [1, 5, 10, 25]
amount = 36
result = coin_change(coins, amount)
print(f"min coins: {result}")
